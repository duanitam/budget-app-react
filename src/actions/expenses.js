import uuid from "uuid";
import database from '../firebase/firebase'
// ------- Action generators - Expenses -------

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState)=>{
        const uid = getState().auth.uid;
        const {  description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
        const expense = { description, note, amount, createdAt};

        // Push the data in firebase database.
        // Return a promise so we can chain it -----
        return database.ref(`users/${uid}/expenses`).push(expense)
            .then( (ref) => {
            dispatch(addExpense({
                id:ref.key, ...expense})) })
            .catch((e) => {console.log(e)})
    }
};

export const removeExpense = ( { id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const editExpense = (id, update) => ({
    type:'EDIT_EXPENSE',
    id,
    update
});

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});


export const startSetExpenses = () => {

    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const expenses = [];

        return database.ref(`users/${uid}/expenses`).once('value')
            .then ((snapshot) => {
            snapshot.forEach( (item) => {
                expenses.push({id:item.key , ...item.val()})});
            dispatch(setExpenses(expenses));
        });
    }
};

export const startRemoveExpense = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense({id}));
        })
    }
};

export const startEditExpense = (id, update) => {
    return (dispatch,getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(update)
            .then( () =>{
                dispatch(editExpense(id,update));
            })
    }
};