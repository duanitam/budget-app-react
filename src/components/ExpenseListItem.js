import React from 'react';
import { Link } from 'react-router-dom'
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = (props) => {
    return(
        <Link className='list-item' to={`/edit/${props.id}`}>

            <div>
                <h3 className='list-item__title'>{props.description}</h3>
                <span className='list-item__subtitle'>{moment(props.createdAt).format('MMMM Do,YYYY')}</span>
            </div>
            <div className='list-item__data'>
                {numeral((props.amount)).format('$0,0.00')}
            </div>
        </Link>
    );
};

export default ExpenseListItem;