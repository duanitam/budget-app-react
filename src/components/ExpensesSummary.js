import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpensesTotal from '../selectors/expenses-total';
import getVisibleExpenses from '../selectors/expenses';
import React from 'react';

export const ExpenseSummary = ( { count, total} ) => {

    return (
        <div>
            { count === 0 ?
                <div> You have no expenses </div> :
                <div> You have {count === 1 ? '1 Expense' : count + ' Expenses' } in summary of {numeral(total/100).format('$0,0.00')} </div>}
        </div>)
    }
;


const mapStateToProps = (state) => ({
    expenses: selectExpensesTotal(getVisibleExpenses(state.expenses, state.filters)),
    count: getVisibleExpenses(state.expenses, state.filters).length
});

export default connect(mapStateToProps)(ExpenseSummary);

