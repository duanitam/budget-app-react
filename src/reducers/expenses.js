
// Expenses Reducer
const expensesReducerDefaultState = [];
const expensesReducer = (state=expensesReducerDefaultState, action) => {

    switch (action.type){
        case 'ADD_EXPENSE':
            return ([...state, action.expense]);

        case 'REMOVE_EXPENSE':
            return state.filter(item => item.id !== action.id);

        case 'EDIT_EXPENSE':
            return state.map(item => (item.id === action.id) ? ({...item,...action.update}): (item));
        default:
            return state
    }
};

export default expensesReducer;