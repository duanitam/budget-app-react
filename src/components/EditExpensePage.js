import React from "react";
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {removeExpense, editExpense} from "../actions/expenses";




export class EditExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/')
    };

    onRemove = () => {
        this.props.removeExpense(this.props.expense.id);
        this.props.history.push('/')
    };

    render() {
        return (<div>
            <ExpenseForm
                expense = {this.props.expense}
                onSubmit={this.onSubmit}/>
            <br/>
            <button onClick={this.onRemove}>Remove</button><br/>
        </div>);
    }
}

const mapStateToProps = (state, props) =>
    ({ expense: state.expenses.find( expense => expense.id === props.match.params.id )});

const mapDispatchToProps = (dispatch, props) => ({
        removeExpense: (data) => dispatch(removeExpense(data)),
        editExpense: (id, expense) => dispatch(editExpense(id,expense))
    });

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);