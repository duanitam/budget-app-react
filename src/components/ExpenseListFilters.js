import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setEndDate, setStartDate } from '../actions/filters'
import {DateRangePicker} from "react-dates";

export class ExpenseListFilters extends React.Component {
    state = {
        calenderFocused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calenderFocused) => {
        this.setState(() => ({calenderFocused}))
    };

    onTextChange = (e) => {
        this.props.setTextFilter({text:e.target.value})};

    onSortChange = (e)=>{
        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();
    };


    render() {
        return (
            <div className='content-container'>
                <div className='input-group'>
                    <div className='input-group__item'>
                        <input className='text_input' placeholder='Search Expenses' type="text" value ={this.props.filters.text} onChange={this.onTextChange}/>
                    </div>
                    <div className='input-group__item'>
                        <select className='select' value={this.props.filters.sortBy} onChange={this.onSortChange}>
                            <option value='date'>Date</option>
                            <option value='amount'>Amount</option>
                        </select></div>
                    <div className='input-group__item'>
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calenderFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={()=> false}
                        />
                    </div>
                </div>


            </div>
        );
    }

}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => {
    return{
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByAmount:() => dispatch(sortByAmount()),
        sortByDate: () => dispatch(sortByDate()),
        setEndDate: (endDate) =>dispatch(setEndDate(endDate)),
        setStartDate: (startDate) => dispatch(setStartDate(startDate))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);