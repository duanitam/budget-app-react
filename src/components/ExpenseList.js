import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div>
        {props.expenses.length === 0 ?
            <p>No expenses</p> :
            props.expenses.map(item => <ExpenseListItem key={item.id} {...item}/>)}
    </div>
);


// Take the state and make it props. get passed to the component
// Things that we get from the store

const mapStateToProps =  (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)     // Orginizing the list according to the filters
    }
};


// Wrapper that connect the component to the store, passing props
export default connect(mapStateToProps)(ExpenseList);
