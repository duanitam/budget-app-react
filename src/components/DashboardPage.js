import React from "react";
import ExpenseList from './ExpenseList'
import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseSummary from './ExpensesSummary'

const ExpensePage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpensePage;