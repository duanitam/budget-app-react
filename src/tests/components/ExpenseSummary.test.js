import  { ExpenseSummary }  from "../../components/ExpensesSummary";
import { shallow } from 'enzyme';
import React from 'react';
import { expenses } from '../fixtures/expenses';


beforeEach( ()=> {

});

test('Should render ExpenseSummary with one item', () => {
    const wrapper = shallow(<ExpenseSummary count={1} total={1241224} />);
    expect(wrapper).toMatchSnapshot();
});

test('Should render ExpenseSummary with multiple items', () => {
    const wrapper = shallow(<ExpenseSummary count={4} total={2342342342}/>);
    expect(wrapper).toMatchSnapshot();

});