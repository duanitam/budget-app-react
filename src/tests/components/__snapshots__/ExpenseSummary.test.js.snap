// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseSummary with multiple items 1`] = `
<div>
  <div>
     You have 
    4 Expenses
     in summary of 
    $23,423,423.42
     
  </div>
</div>
`;

exports[`Should render ExpenseSummary with one item 1`] = `
<div>
  <div>
     You have 
    1 Expense
     in summary of 
    $12,412.24
     
  </div>
</div>
`;
